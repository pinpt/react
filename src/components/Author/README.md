### An author name and avatar

[Full Docs](https://react.preview.pinpoint.com/?path=/docs/components-avatar) | [Github](https://github.com/pinpt/react/tree/master/src/components/Avatar)

#### Theming

The parent className for styling this component is `.Pinpoint.Author`

<details>
	<summary>Base Theme Styles</summary>

```css
.Pinpoint.Author,
.Pinpoint.Author a {
	@apply flex items-center mr-auto;
}

.Pinpoint.Author .avatar {
	@apply rounded-full mr-2 w-8 h-8 border-2;
	border-color: var(--page-secondary-text-color);
}

.Pinpoint.Author .name {
	@apply whitespace-nowrap;
}
```

</details>
