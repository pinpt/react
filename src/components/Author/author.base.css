.Pinpoint.Author,
.Pinpoint.Author > .Pinpoint.Link,
.Pinpoint.Author a {
	@apply flex items-center mr-auto;
}

.Pinpoint.Author .avatar {
	@apply rounded-full mr-2 w-8 h-8 border-2;
	border-color: var(--page-secondary-text-color);
}

.Pinpoint.Author .name {
	@apply whitespace-nowrap;
}
