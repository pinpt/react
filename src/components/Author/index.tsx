export interface IAuthorProps {
	className?: string;
	avatarUrl: string;
	name?: string;
}

const Author = (props: IAuthorProps) => {
	const { className = '', avatarUrl, name } = props;

	return (
		<div className={`Pinpoint Author ${className}`}>
			<img className="avatar" src={avatarUrl} />
			<span className="name">{name}</span>
		</div>
	);
};

export default Author;
