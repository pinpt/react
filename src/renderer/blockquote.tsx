import { registerNode, recurseIntoChildren, NodeProps } from './register';

const BlockQuote = ({ node }: NodeProps) => {
	return <blockquote>{recurseIntoChildren(node)}</blockquote>;
};

registerNode('blockquote', (node) => <BlockQuote node={node} />);
