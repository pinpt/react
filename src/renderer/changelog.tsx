import { ReactYouTubeLite as Youtube } from 'react-youtube-lite';
import { getFileUrl } from '../lib/file_metadata';
import { slugifyContent } from '../lib/string';
import { ChangelogCoverMedia, ChangelogMediaType } from '../lib/types';
import { Content as Doc } from './';

const ImageMedia = ({ src, title = '' }: { src: string; title?: string }) => {
	return (
		<div className="image">
			<img src={src} alt={title} />
		</div>
	);
};

const VideoMedia = ({ src }: { src: string }) => {
	return (
		<div className="video">
			<video controls src={src} />
		</div>
	);
};

const YoutubeMediaSSR = ({ id, poster = 'hqdefault' }: { id: string; poster?: string }) => {
	// the component isn't SSR friendly so this is basically used to sort of mimic it for generated code
	const posterUrl = `https://i.ytimg.com/vi/${id}/${poster}.jpg`;
	return (
		<>
			{poster && <link rel="preload" href={posterUrl} as="image" />}
			<div
				className="youtube ryt-lite embed-responsive aspect-ratio-16/9"
				style={{ backgroundImage: `url(${posterUrl})` }}
			>
				<div className="lty-playbtn" data-url={`https://www.youtube-nocookie.com/embed/${id}?autoplay=1`} />
			</div>
		</>
	);
};

const YoutubeMedia = ({
	id,
	metadata,
	staticMode,
}: {
	id: string;
	metadata?: Record<string, any>;
	staticMode?: true;
}) => {
	return (
		<div className="youtube">
			{!staticMode ? (
				<Youtube url={`https://www.youtube.com/watch?v=${id}`} poster={metadata?.poster} />
			) : (
				<YoutubeMediaSSR id={id} poster={metadata?.poster} />
			)}
		</div>
	);
};

const Media = ({ media, title, staticMode }: { media?: ChangelogCoverMedia; title?: string; staticMode?: true }) => {
	if (!media) {
		return <></>;
	}
	let content: React.ReactElement;
	switch (media.type) {
		case ChangelogMediaType.None: {
			return <></>;
		}
		case ChangelogMediaType.Image: {
			content = <ImageMedia src={getFileUrl(media.value)} title={title} />;
			break;
		}
		case ChangelogMediaType.Video: {
			content = <VideoMedia src={getFileUrl(media.value)} />;
			break;
		}
		case ChangelogMediaType.Youtube: {
			content = <YoutubeMedia id={media.value} metadata={media.metadata} staticMode={staticMode} />;
			break;
		}
	}
	return <div className="media-container">{content}</div>;
};

interface ChangelogProps {
	id: string;
	title?: string;
	content?: any;
	coverMedia?: ChangelogCoverMedia;
	limit?: number;
	divider?: boolean;
	staticMode?: true;
}

const ChangelogLog = (props: ChangelogProps) => {
	return (
		<article className="changelog notebook-editor read-only">
			{props.coverMedia && (
				<section className="covermedia">
					<Media media={props.coverMedia} title={props.title} staticMode={props.staticMode} />
				</section>
			)}
			<section className="ProseMirror">
				<Doc node={props.content} limit={props.limit} />
				{props.limit && props.content.content?.length > props.limit && (
					<div className="continue">
						<a href={slugifyContent(props.id, props.title)}>Continue Reading â†’</a>
					</div>
				)}
			</section>
			{props.divider && <hr className="divider" />}
		</article>
	);
};

export default ChangelogLog;
