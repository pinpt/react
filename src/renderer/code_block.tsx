import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter';
import { twilight } from 'react-syntax-highlighter/dist/cjs/styles/prism';
import { NodeProps, registerNode } from './register';

const CodeBlock = ({ node }: NodeProps) => {
	return (
		<SyntaxHighlighter
			language={node.attrs.language}
			style={twilight}
			customStyle={{
				background: 'black',
				fontFamily: 'ui-monospace',
				borderWidth: '0px',
				borderRadius: '.25em',
			}}
		>
			{node.content?.[0].text ?? ''}
		</SyntaxHighlighter>
	);
};

registerNode('code_block', (node) => <CodeBlock node={node} />);
