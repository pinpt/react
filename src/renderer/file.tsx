import {
	faDownload, faFile, faFileAlt, faFileArchive, faFileAudio, faFileCode, faFileExcel, faFilePdf,
	faFilePowerpoint, faFileVideo, faFileWord
} from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { NodeProps, registerNode } from './register';

export const iconForType = (type = '') => {
	if (/pdf/.test(type)) {
		return faFilePdf;
	}
	if (/zip|tar/.test(type)) {
		return faFileArchive;
	}
	if (/video/.test(type)) {
		return faFileVideo;
	}
	if (/audio/.test(type)) {
		return faFileAudio;
	}
	if (/excel|spreadsheet/.test(type)) {
		return faFileExcel;
	}
	if (/powerpoint/.test(type)) {
		return faFilePowerpoint;
	}
	if (/word/.test(type)) {
		return faFileWord;
	}
	if (/css|csv|graphql|html|java|json|php|python|typescript|xml/.test(type)) {
		return faFileCode;
	}
	if (/text/.test(type)) {
		return faFileAlt;
	}
	return faFile;
};

export const formattedSize = (size = 0) => {
	const kB = size / 1000;
	if (kB < 10) {
		return `${Math.round(kB * 10) / 10} KB`;
	}
	if (kB < 1000) {
		return `${Math.round(kB)} KB`;
	}
	const MB = kB / 1000;
	return `${Math.round(MB * 10) / 10} MB`;
};

const File = ({ node }: NodeProps) => {
	const { description, filename, publicUrl, size, type } = node.attrs;

	const onDownload = () => {
		if (publicUrl) {
			const u = new URL(publicUrl);
			u.searchParams.set('filename', filename);
			u.searchParams.set('disposition', 'attachment');
			const a = document.createElement('a');
			a.style.display = 'none';
			a.href = u.toString();
			// the filename you want
			a.download = filename;
			document.body.appendChild(a);
			a.click();
			a.remove();
		}
	};

	return (
		<div className="file">
			<div className="container">
				<FontAwesomeIcon icon={iconForType(type)} />
				<div className="details">
					{publicUrl ? (
						<a href={publicUrl} target="_blank" rel="noopener noreferrer" className="filename">
							{filename}
						</a>
					) : (
						<div className="filename">{filename}</div>
					)}
					{description && <div className="description">{description}</div>}
					<div className="size">{formattedSize(size)}</div>
				</div>
				<div className="options">
					{publicUrl && (
						<div className="option group" onClick={onDownload}>
							<FontAwesomeIcon icon={faDownload} />
							<span>Download</span>
						</div>
					)}
				</div>
			</div>
		</div>
	);
};

registerNode('file', (node) => <File node={node} />);
