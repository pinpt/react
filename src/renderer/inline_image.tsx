import { extractImageMetadataFromFileID } from '../lib/file_metadata';
import { NodeProps, registerNode } from './register';

const InlineImage = ({ node }: NodeProps) => {
	const { size } = extractImageMetadataFromFileID(node.attrs?.src);
	const width = size?.width;
	const height = size?.height;

	const image = (
		<img src={node.attrs?.src} width={width} height={height} alt={node.attrs?.alt} className="inline_image" />
	);

	return (
		<div>
			{node.attrs?.href ? (
				<a href={node.attrs.href} target="_blank" rel="noopener noreferrer">
					{image}
				</a>
			) : (
				<>{image}</>
			)}
		</div>
	);
};

registerNode('inline-image', (node) => <InlineImage node={node} />);
