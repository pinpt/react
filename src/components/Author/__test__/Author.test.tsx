import renderer from 'react-test-renderer';
import Author from '..';

const AVATAR =
	'https://file.pinpoint.com/6cc744395c563f0072b339070aa327f3;UCGt%3D%3DkE~A%3FZJ6%5EhIpS%23o%5BELE3oLT0I%40t7E2;400x400.png?rw=48&rh=48';

test('Test default', () => {
	const component = renderer.create(<Author avatarUrl={AVATAR} name="Keegan" />);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Test no name', () => {
	const component = renderer.create(<Author avatarUrl={AVATAR} />);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Test custom className', () => {
	const component = renderer.create(<Author avatarUrl={AVATAR} name="Keegan" className="custom-test" />);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});
