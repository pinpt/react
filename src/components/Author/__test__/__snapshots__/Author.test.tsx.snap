// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test custom className 1`] = `
<div
  className="Pinpoint Author custom-test"
>
  <span
    aria-label="Keegan avatar"
    className="Pinpoint Internal Link  "
  >
    <img
      alt=""
      className="Pinpoint Image avatar "
      decoding="async"
      height={48}
      loading="lazy"
      src="data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw=="
      width={48}
    />
    <span
      className="name"
    >
      Keegan
    </span>
  </span>
</div>
`;

exports[`Test default 1`] = `
<div
  className="Pinpoint Author "
>
  <span
    aria-label="Keegan avatar"
    className="Pinpoint Internal Link  "
  >
    <img
      alt=""
      className="Pinpoint Image avatar "
      decoding="async"
      height={48}
      loading="lazy"
      src="data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw=="
      width={48}
    />
    <span
      className="name"
    >
      Keegan
    </span>
  </span>
</div>
`;

exports[`Test no name 1`] = `
<div
  className="Pinpoint Author "
>
  <span
    aria-label=" avatar"
    className="Pinpoint Internal Link  "
  >
    <img
      alt=""
      className="Pinpoint Image avatar "
      decoding="async"
      height={48}
      loading="lazy"
      src="data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw=="
      width={48}
    />
    <span
      className="name"
    >
      
    </span>
  </span>
</div>
`;
